# **GAS STORAGE OPTIMIZATION**

In [ ]:
import sys
sys.path.append(r"C:\Users\Elie\Elie code\Function\TTF")
import warnings
warnings.filterwarnings('ignore')
# import des fonctions
from Storage_Optimization import StorageOptimization

**Function import**

In [ ]:
StorageOptimization = StorageOptimization()

**Load the df with all the forward contract price (Oct 25 -> Mars 27) since 2025**

In [ ]:
TTF = StorageOptimization.get_contracts_dataframe(market='TTF', contract_type='monthly')
TTF = TTF[:-1]
TTF.loc['2025-03-07']

Chargement des données daily depuis C:\Users\Elie\Datas\TradingView\TTF curve...
37 fichiers CSV trouvés


TTFOct25  TTFNov25  TTFDec25  TTFJan26  TTFFeb26  \
Date                                                                    
2025-01-02 07:00:00    47.653    46.243    45.928    45.600    44.835   
2025-01-03 07:00:00    47.199    46.000    45.720    45.515    44.795   
2025-01-06 07:00:00    45.760    44.655    44.380    44.150    43.462   
2025-01-07 07:00:00    46.042    44.962    44.697    44.475    43.835   
2025-01-08 07:00:00    44.462    43.432    43.237    43.117    42.542   
...                       ...       ...       ...       ...       ...   
2025-10-09 06:00:00    32.060    32.363    32.679    32.899    32.889   
2025-10-10 06:00:00    32.060    32.170    32.428    32.668    32.689   
2025-10-13 06:00:00    32.060    31.475    31.753    32.018    32.058   
2025-10-14 06:00:00    32.060    31.801    32.033    32.258    32.283   
2025-10-15 06:00:00    32.060    31.826    32.024    32.216    32.253   

                     TTFMar26  TTFApr26  TTFMay26  TTFJun26  TTFJul26  \
Date                                                                    
2025-01-02 07:00:00    43.420    39.506    38.551    38.074    37.828   
2025-01-03 07:00:00    43.430    39.727    38.759    38.315    38.108   
2025-01-06 07:00:00    42.187    39.104    38.169    37.724    37.554   
2025-01-07 07:00:00    42.540    39.314    38.376    37.971    37.821   
2025-01-08 07:00:00    41.317    38.266    37.401    36.996    36.784   
...                       ...       ...       ...       ...       ...   
2025-10-09 06:00:00    32.421    31.194    30.739    30.694    30.737   
2025-10-10 06:00:00    32.189    30.965    30.515    30.456    30.501   
2025-10-13 06:00:00    31.599    30.468    30.035    29.985    30.058   
2025-10-14 06:00:00    31.834    30.634    30.162    30.117    30.152   
2025-10-15 06:00:00    31.813    30.616    30.162    30.116    30.148   

                     TTFAug26  TTFSep26  TTFOct26  TTFNov26  TTFDec26  \
Date                                                                    
2025-01-02 07:00:00    37.784    37.934    37.454    37.299    37.399   
2025-01-03 07:00:00    38.063    38.150    37.650    37.367    37.477   
2025-01-06 07:00:00    37.509    37.659    37.199    36.964    37.069   
2025-01-07 07:00:00    37.777    37.929    37.423    37.183    37.293   
2025-01-08 07:00:00    36.744    36.899    36.584    36.351    36.505   
...                       ...       ...       ...       ...       ...   
2025-10-09 06:00:00    30.852    31.060    31.247    31.958    32.370   
2025-10-10 06:00:00    30.621    30.821    31.006    31.732    32.114   
2025-10-13 06:00:00    30.179    30.384    30.582    31.273    31.653   
2025-10-14 06:00:00    30.257    30.477    30.659    31.354    31.734   
2025-10-15 06:00:00    30.261    30.466    30.629    31.298    31.704   

                     TTFJan27  TTFFeb27  TTFMar27  
Date                                               
2025-01-02 07:00:00    37.206    36.561    35.316  
2025-01-03 07:00:00    37.347    36.707    35.463  
2025-01-06 07:00:00    36.924    36.284    35.044  
2025-01-07 07:00:00    37.118    36.478    35.238  
2025-01-08 07:00:00    36.593    35.948    34.383  
...                       ...       ...       ...  
2025-10-09 06:00:00    32.491    32.361    31.621  
2025-10-10 06:00:00    32.215    32.127    31.352  
2025-10-13 06:00:00    31.793    31.693    30.938  
2025-10-14 06:00:00    31.872    31.767    31.009  
2025-10-15 06:00:00    31.855    31.720    30.971  

[203 rows x 18 columns]

**Stores the df in the StorageOptimization for processing**

In [ ]:
StorageOptimization.load_forward_curve_data(TTF)  

=== LOADING FORWARD CURVE DATA ===
DataFrame shape: (203, 18)
Date range: 2025-01-02 07:00:00 to 2025-10-15 06:00:00
Contracts loaded: ['TTFOct25', 'TTFNov25', 'TTFDec25', 'TTFJan26', 'TTFFeb26', 'TTFMar26', 'TTFApr26', 'TTFMay26', 'TTFJun26', 'TTFJul26', 'TTFAug26', 'TTFSep26', 'TTFOct26', 'TTFNov26', 'TTFDec26', 'TTFJan27', 'TTFFeb27', 'TTFMar27']
Total data points: 3654
Non-NaN values: 3654 (100.0%)

Current forward curve (2025-10-15):
  TTFOct25: 32.060 €/MWh
  TTFNov25: 31.826 €/MWh
  TTFDec25: 32.024 €/MWh
  TTFJan26: 32.216 €/MWh
  TTFFeb26: 32.253 €/MWh
  TTFMar26: 31.813 €/MWh
  TTFApr26: 30.616 €/MWh
  TTFMay26: 30.162 €/MWh
  TTFJun26: 30.116 €/MWh
  TTFJul26: 30.148 €/MWh
  TTFAug26: 30.261 €/MWh
  TTFSep26: 30.466 €/MWh
  TTFOct26: 30.629 €/MWh
  TTFNov26: 31.298 €/MWh
  TTFDec26: 31.704 €/MWh
  TTFJan27: 31.855 €/MWh
  TTFFeb27: 31.720 €/MWh
  TTFMar27: 30.971 €/MWh


# **STEP 1: Set storage parameters first**

In [ ]:
StorageOptimization.set_storage_parameters(
    capacity_gwh=175.8,
    max_injection_gwh_per_day=5.86,
    max_withdrawal_gwh_per_day=5.86,
    fixed_cost_euros=200000
)

=== 💰 STORAGE COST BREAKDOWN ===
🏠 Fixed Cost: 0.0031 €/MWh/day
⬇️  Injection Cost: 0.1 €/MWh
⬆️  Withdrawal Cost: 0.1 €/MWh
📊 Total Variable Cost: 0.2 €/MWh

=== 🧮 COST CALCULATION FORMULA ===
Fixed Cost per Day = fixed_cost_euros/ 365 days
Fixed Cost per MWh = Fixed Cost per Day / Total Capacity in MWh
175.8 GWh capacity = 175,800 MWh
Each MWh of storage capacity costs €0.0031 per day


# **STEP 2: Launch the Backtest**

In [ ]:
# Appeler la fonction et récupérer tous les résultats
results = StorageOptimization.calculate_rolling_intrinsic_value_with_costs()  


=== ROLLING INTRINSIC VALUE - WITH REAL SALES ===
💰 COSTS: Injection=0.1€/MWh, Withdrawal=0.1€/MWh, Fixed=200,000€
📅 PERIODS: - Trading=2025-01-01 to 2026-04-01
             - Storage Allocation=2026-04-01 to 2027-03-31

📊 ROLLING INTRINSIC STRATEGY - WITH REAL SALES

🎯 OVERALL PERFORMANCE:
    • Initial Value: 0 €
    • Trading Period (2025-01-01 to 2026-04-01):
       - Rebalances: 73
       - Profit: 367,070 €
    • Storage Period (2026-04-01 to 2027-03-31):
       - Rebalances: 0
       - Operational Profit: 0 €
       - Fixed Cost: -200,000 €
    • Final Intrinsic Value: 274,248 €
    • TOTAL NET PROFIT: 167,070 €

📈 REBALANCING SUMMARY:
Date         Old Val    New Val    Realized PnL Net Benefit 
------------------------------------------------------------
2025-01-08           0€     7,384€           0€       7,384€
2025-01-09       7,384€         0€      12,482€       5,098€
2025-02-12           0€    43,071€       3,516€      46,587€
2025-02-19      86,142€    79,110€      14,

**A. Current Actual Performance (From Your Backtest)**

from simulation:
- Rebalancing Profit (10 months): 367,070 €
- Number of Trades: 73
- Fixed Cost (annual): 200,000 €

Simple math:
- Net P&L so far = 367,070 € - 200,000 €
                 = **167,070 €**

**C. Projection to Full Year**

I start trading in Jan25.
If we continue at same pace for 2 more months (until Dec25) :

- Additional profit = 367,070 € / (10) = 36,707 €
- Total annual profit = 36,707 € * 12 = 440 484 €
- Annual net P&L = 440 484 € - €200,000 € = **240 484 €**